# UTS-Regression dengan KNN
Disini saya Load Boston Housing Price yang disertakan sebagai sample dataset pada Scikit Learn.

In [1]:
from sklearn.datasets import load_boston

boston = load_boston()
boston

{'data': array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
         4.9800e+00],
        [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
         9.1400e+00],
        [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
         4.0300e+00],
        ...,
        [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         5.6400e+00],
        [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
         6.4800e+00],
        [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         7.8800e+00]]),
 'target': array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
        18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
        15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
        13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
        21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
        35.4, 24.7, 3

In [2]:
print(boston.DESCR)

.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pu

In [3]:
from sklearn import datasets
import pandas as pd

boston_data = datasets.load_boston()
df_boston = pd.DataFrame(boston_data.data,columns=boston_data.feature_names)
df_boston['target'] = pd.Series(boston_data.target)
df_boston.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  target  
0     15.3  396.90   4.98    24.0  
1     17.8  396.90   9.14    21.6  
2     17.8  392.83   4.03    34.7  
3     18.7  394.63   2.94    33.4  
4     18.7  396.90   5.33    36.2

## Sample Dataset
Disini saya mengambil 2 features yaitu RM dan RAD. RM sebagai jumlah rata-rata kamar per hunian sedangkan RAD yaitu indeks aksesibilitas ke jalan raya radial. Saya mengambil features RM dan RAD karena saya ingin mengetahui prediksi harga rumah bila diketahui jumlah rata-rata kamar per hunian dan indeks aksesibilitas ke jalan raya radial.
- Rata-rata kamar dan index akses sebagai features.
- Harga sebagai target.

In [4]:
import pandas as pd

rumah = {'rata-rata kamar': [6.575, 6.421, 7.185, 6.998, 7.147], 
          'index akses': [1, 2, 2, 3, 3],
          'harga': [24, 21.6, 34.7, 33.4, 36.2]}

rumah_df = pd.DataFrame(rumah)
rumah_df

rata-rata kamar  index akses  harga
0            6.575            1   24.0
1            6.421            2   21.6
2            7.185            2   34.7
3            6.998            3   33.4
4            7.147            3   36.2

## Regression (Estimasi Nilai) dengan KNN
### Features & Target

In [5]:
import numpy as np

X_train = np.array(rumah_df[['rata-rata kamar', 'index akses']])
y_train = np.array(rumah_df['harga'])

print(f'X_train:\n{X_train}\n')
print(f'y_train: {y_train}')

X_train:
[[6.575 1.   ]
 [6.421 2.   ]
 [7.185 2.   ]
 [6.998 3.   ]
 [7.147 3.   ]]

y_train: [24.  21.6 34.7 33.4 36.2]


### Training KNN Regression Model
Mengimport KNeighborsRegressor karena akan menggunakan KNN untuk regression tasks.

In [6]:
from sklearn.neighbors import KNeighborsRegressor

K = 3
model = KNeighborsRegressor(n_neighbors=K)
model.fit(X_train, y_train)

KNeighborsRegressor(n_neighbors=3)

### Prediksi Harga Rumah

In [7]:
X_new = np.array([[7.185, 2]])
X_new

array([[7.185, 2.   ]])

In [8]:
y_pred = model.predict(X_new)
y_pred

array([30.83333333])

Diprediksi dengan rata-rata kamar 7.185 dan index akses 2, memiliki harga 30.83333333.

### Evaluasi KNN Regression Model

In [9]:
X_test = np.array([[6.595, 1], [7.543, 1], [7.235, 2], [9.549, 3], [8.436, 3]])
y_test = np.array([27.3, 28.5, 32.7, 33.1, 35.9])

print(f'X_test:\n{X_test}\n')
print(f'y_test: {y_test}')

X_test:
[[6.595 1.   ]
 [7.543 1.   ]
 [7.235 2.   ]
 [9.549 3.   ]
 [8.436 3.   ]]

y_test: [27.3 28.5 32.7 33.1 35.9]


In [10]:
y_pred = model.predict(X_test)
y_pred

array([26.76666667, 26.76666667, 30.83333333, 34.76666667, 34.76666667])

#### Coefficient of Determination atau $R^2$

In [11]:
from sklearn.metrics import r2_score

r_squared = r2_score(y_test, y_pred)

print(f'R-squared: {r_squared}')

R-squared: 0.783288888888889


R-squared semakin mendekati 1 maka semakin baik dan semakin mendekati 0 maka semakin buruk pula. Tidak menutup kemungkinan mendapatkan nilai (-) sebagai nilai terburuk.

#### Mean Absolute Error (MAE) atau Mean Absolute Deviation (MAD)

In [12]:
from sklearn.metrics import mean_absolute_error

MAE = mean_absolute_error(y_test, y_pred)

print(f'MAE: {MAE}')

MAE: 1.386666666666666


Semakin kecil nilai MAE akan mengindikasikan kualitas model yang lebih baik.

#### Mean Squared Error (MSE) atau Mean Squared Deviation (MSD)

In [13]:
from sklearn.metrics import mean_squared_error

MSE = mean_squared_error(y_test, y_pred)

print(f'MSE: {MSE}')

MSE: 2.16711111111111


Semakin kecil nilai MSE akan mengindikasikan kualitas model yang lebih baik.